### PR TITLE
Visitors validation API

### PR DESCRIPTION
**Pull request by alan**
Takes care of validation the visitors for a visit:
- Lead visitor age
- Number of adults (prison defined age)
- Number of visitors

### CODE REVIEW

**Comment by alan** [comment]
Related to https://github.com/ministryofjustice/prison-visits-public/pull/54

File: app/models/visitors_validation.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  def lead_visitor_age
+    if age_calculator.age(lead_date_of_birth) < MIN_LEAD_VISITOR_AGE
+      errors.add(:general, 'lead_visitor_age')
+    end
+  end
+
+  def number_of_visitors
+    if dates_of_birth.size > Prison::MAX_VISITORS
+      errors.add(:general, 'too_many_visitors')
```
Comment (alan) [inline app/models/visitors_validation.rb]: Using strings matters a lot. When using symbols the error message is localised when fetching the values on `errors`, had me confused for a while.

Comment (mloughran) [inline app/models/visitors_validation.rb]: Gotta love the magic...

File: spec/models/visitors_validation_spec.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        subject.valid?
+        expect(subject.error_keys).to eq(['too_many_visitors'])
+      end
+    end
+
+    context 'when there are too many adults' do
+      let(:dobs) do
+        (Prison::MAX_ADULTS + 1).times.map {
+          prison.adult_age.years.ago.to_date
+        }
```
Comment (mloughran) [inline spec/models/visitors_validation_spec.rb]: I wonder if it's clearer to write this? (ditto elsewhere)

    [prison.adult_age.years.ago.to_date] * (Prison::MAX_ADULTS + 1)

**Comment by mloughran** [comment]
@alan what are your thoughts on the duplication of visitor validation between this code and e.g. `Prison#validate_visitor_ages_on`?

**Comment by alan** [comment]
I didn't know of that method. I think the duplication, when we remove the steps code then we can delete `Prison#validate_visitor_ages_on`

**Comment by mloughran** [comment]
@alan +1 for refactoring when we remove the steps. I think it's a bit more nuanced – the request booking method still needs to validate the booking (it can't rely on this validation API having been called) – but we'll find a solution.

File: app/models/visitors_validation.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
+class VisitorsValidation
+  include NonPersistedModel
+
+  MIN_LEAD_VISITOR_AGE = 18 # years
```
Comment (mloughran) [inline app/models/visitors_validation.rb]: This is duplicated in Prison also. You're using `Prison::MAX_VISITORS` so probably best to be consistent.

Comment (alan) [inline app/models/visitors_validation.rb]: Good catch

**Comment by alan** [comment]
Yes, I expect the request booking method (I guess you mean to create a visit) will run the same set of validations as the validations endpoints as a minimum.

**Pull request merged**